cf-custom-resources/lib/dns-cert-validator.js

```diff
@@ -161,19 +192,197 @@ const requestCertificate = async function (
+  domainName,
+  envRoute53,
+  appRoute53,
+  envHostedZoneId
+) {
+  const promises = [];
+  for (const option of options) {
+    switch (option.DomainName) {
+      case `${envName}.${appName}.${domainName}`:
+        promises.push(
+          validateDomain({
+            route53: envRoute53,
+            record: option.ResourceRecord,
+            action: action,
+            domainName: "",
+            hostedZoneId: envHostedZoneId,
+          })
+        );
+        break;
+      case `${appName}.${domainName}`:
+        promises.push(
+          validateDomain({
+            route53: appRoute53,
+            record: option.ResourceRecord,
+            action: action,
+            domainName: `${appName}.${domainName}`,
+          })
+        );
+        break;
+      case domainName:
+        promises.push(
+          validateDomain({
+            route53: appRoute53,
+            record: option.ResourceRecord,
+            action: action,
+            domainName: domainName,
+          })
+        );
+        break;
+    }
+  }
+  return Promise.all(promises);
+};
+
+const deleteHostedZoneRecords = async function (
+  options,
+  envName,
+  appName,
+  domainName,
+  envRoute53,
+  appRoute53,
+  acm,
+  envHostedZoneId
+) {
+  let isLegacyCert = false;
+  // Legacy cert only has two DomainValidationOptions.
+  if (options.length <= 2) {
+    isLegacyCert = true;
+  }
+
+  let isLastOne = false;
+  let certWithDomainCount = 0;
+  let listCertResp = await acm.listCertificates({}).promise();
+  for (const certSummary of listCertResp.CertificateSummaryList || []) {
+    if (certSummary.DomainName === `${envName}.${appName}.${domainName}`) {
+      certWithDomainCount++;
+    }
+  }
+  let nextToken = listCertResp.NextToken;
+  // Though very unlikely, we'll handle the pagination.
+  while (nextToken && certWithDomainCount < 2) {
+    listCertResp = await acm
+      .listCertificates({
+        NextToken: nextToken,
+      })
+      .promise();
+    for (const certSummary of listCertResp.CertificateSummaryList || []) {
+      if (certSummary.DomainName === `${envName}.${appName}.${domainName}`) {
+        certWithDomainCount++;
+      }
+    }
+    nextToken = listCertResp.NextToken;
+  }
+  if (certWithDomainCount < 2) {
+    isLastOne = true;
+  }
+
+  const newOptions = [];
+  switch (`${isLegacyCert}|${isLastOne}`) {
+    case `true|true`:
+      // If it is a legacy cert and it is the last Copilot cert,
+      // we'll go ahead to remove the validation CNAME record in env hosted zone.
+      newOptions.push(...options);
+      break;
+    case `true|false`:
+      // If it is a legacy cert but it is not the last Copilot cert,
+      // we'll do nothing since the new Copilot cert needs validation CNAME record
+      // in the env hosted zone.
+      break;
+    case `false|true`:
+      // If it is not a legacy cert and it is not the last Copilot cert,
+      // we'll remove all validation CNAME records in env/app/root hosted zone.
```

Comment:
Isn't this comment for the `false|false` scenario? I think the logic inside the `case` is correct it's just the comments that need to be swapped.